src/hera/workflows/runner.py

```diff
@@ -122,7 +123,57 @@ def _map_keys(function: Callable, kwargs: dict) -> dict:
     return mapped_kwargs
 
 
-def _runner(entrypoint: str, kwargs_list: Any) -> str:
+def _save_outputs(
+    function_results: Union[Tuple[Any], Any], output_destinations: List[Tuple[type, Union[Parameter, Artifact]]]
+) -> None:
+    """Save the results of the function to the specified output destinations.
+
+    The results are matched with the specified outputs and saved using the schema:
+    <parent_directory>/artifacts/<name>
+    <parent_directory>/parameters/<name>
+    If the artifact path is specified, that is used instead.
+    <parent_directory> can be provided by the user or is set to /hera/outputs by default
+    """
+    if not isinstance(function_results, tuple):
+        function_results = [function_results]
+    if len(function_results) != len(output_destinations):
+        raise ValueError("The number of outputs does not match the annotation")
+
+    for res, dest in zip(function_results, output_destinations):
+        if not isinstance(res, dest[0]):
+            raise ValueError(
+                f"The type of output `{dest[1].name}`, `{type(res)}` does not match the annotated type `{dest[0]}`"
+            )
+        if not dest[1].name:
+            raise ValueError("The name was not provided for one of the outputs.")
+
+        path = _get_outputs_path(dest[1])
+        _write_to_path(path, res)
+
+
+def _get_outputs_path(destination: Union[Parameter, Artifact]) -> Path:
+    """Get the path from the destination annotation using the defined outputs directory."""
+    outputs_dir = Path(os.environ.get("hera__outputs_directory", "hera/outputs"))
+    if isinstance(destination, Parameter):
+        path = outputs_dir / f"parameters/{cast(str, destination.name)}"
+    elif isinstance(destination, Artifact):
+        if destination.path:
+            path = Path(destination.path)
+        else:
+            path = outputs_dir / f"artifacts/{cast(str, destination.name)}"
+    return path
+
+
+def _write_to_path(path: Path, file) -> None:
+    """Write the file contents to the provided path. Create the necessary parent directories."""
+    directory = path.parent
+    directory.mkdir(parents=True, exist_ok=True)
```

Comment:
Nitpicking, but because `directory` isn't used again
```suggestion
    path.parent.mkdir(parents=True, exist_ok=True)
```